File: companion/bolumler/b11/ornek-01/cozum.R
SUTUNLAR <- c("hacim_1", "ortalama_1", "standart_sapma_1", "hacim_2", "ortalama_2",
               "standart_sapma_2", "cift_sayisi", "ortalama_fark", "fark_sapmasi", "alfa")

hazirla <- function(veri) {
  if (!identical(names(veri), SUTUNLAR) || nrow(veri) != 1) stop("Tek ozet satiri ve dogru sutun sirasi gerekli.")
  if (!all(vapply(veri, is.numeric, logical(1))) || anyNA(veri) ||
      any(!is.finite(as.matrix(veri)))) stop("Sonlu sayisal ozet gerekli.")
  hacimler <- unlist(veri[c("hacim_1", "hacim_2", "cift_sayisi")])
  if (any(hacimler < 2 | hacimler != floor(hacimler))) stop("Hacimler en az 2 olan tam sayilar olmali.")
  if (any(unlist(veri[c("standart_sapma_1", "standart_sapma_2", "fark_sapmasi")]) <= 0)) {
    stop("Bu paket pozitif standart sapmalar gerektirir.")
  }
  if (veri$alfa <= 0 || veri$alfa >= 1) stop("Alfa 0 ile 1 arasinda olmali.")
  veri
}

test_ozeti <- function(fark, standart_hata, serbestlik, alfa) {
  t_degeri <- fark / standart_hata
  p_degeri <- 2 * pt(abs(t_degeri), df = serbestlik, lower.tail = FALSE)
  kritik <- qt(1 - alfa / 2, df = serbestlik)
  hata_payi <- kritik * standart_hata
  sonuc <- c(standart_hata = standart_hata, serbestlik = serbestlik, t = t_degeri,
              p_cift = p_degeri, kritik_t = kritik, hata_payi = hata_payi,
              alt_sinir = fark - hata_payi, ust_sinir = fark + hata_payi,
              genislik = 2 * hata_payi, reddet_cift = as.integer(p_degeri < alfa))
  if (any(!is.finite(sonuc))) stop("Test sayisal olarak hesaplanamadi.")
  sonuc
}

hesapla <- function(veri) {
  veri <- hazirla(veri)
  katki_1 <- veri$standart_sapma_1^2 / veri$hacim_1
  katki_2 <- veri$standart_sapma_2^2 / veri$hacim_2
  toplam <- katki_1 + katki_2
  serbestlik <- toplam^2 / (katki_1^2 / (veri$hacim_1 - 1) + katki_2^2 / (veri$hacim_2 - 1))
  fark <- veri$ortalama_1 - veri$ortalama_2
  welch <- c(katki_1 = katki_1, katki_2 = katki_2, fark = fark,
              test_ozeti(fark, sqrt(toplam), serbestlik, veri$alfa))
  eslesmis <- c(test_ozeti(veri$ortalama_fark, veri$fark_sapmasi / sqrt(veri$cift_sayisi),
                          veri$cift_sayisi - 1, veri$alfa), dz = veri$ortalama_fark / veri$fark_sapmasi)
  sonuc <- data.frame(degisken = c(rep("girdi", 10), rep("welch", 13), rep("eslesmis", 11)),
                       olcu = c(names(veri), names(welch), names(eslesmis)),
                       deger = c(unlist(veri), welch, eslesmis), row.names = NULL)
  if (any(!is.finite(sonuc$deger))) stop("Sonlu sonuc hesaplanamadi.")
  sonuc
}

kontrol_et <- function(sonuc, beklenen) {
  if (!identical(names(beklenen), c("degisken", "olcu", "deger")) ||
      !identical(sonuc$degisken, beklenen$degisken) || !identical(sonuc$olcu, beklenen$olcu)) {
    stop("Kontrol etiketleri veya sirasi uyusmuyor.")
  }
  if (!is.numeric(beklenen$deger) || any(!is.finite(beklenen$deger)) ||
      any(!is.finite(sonuc$deger))) stop("Sonlu kontrol degerleri gerekli.")
  tolerans <- pmax(1e-9, 1e-9 * pmax(abs(sonuc$deger), abs(beklenen$deger)))
  if (any(abs(sonuc$deger - beklenen$deger) > tolerans)) stop("Kontrol degerleri uyusmuyor.")
  cat("DOGRULANDI:", nrow(sonuc), "kontrol degeri eslesiyor.\n")
}

grafik_kaydet <- function(veri) {
  veri <- hazirla(veri)
  sonuc <- hesapla(veri)
  dir.create("ciktilar/r", recursive = TRUE, showWarnings = FALSE)
  png("ciktilar/r/fark-araliklari.png", width = 1500, height = 600, res = 150)
  on.exit(dev.off())
  par(mfrow = c(1, 2))
  basliklar <- c(welch = "Welch: Grup 1 - Grup 2", eslesmis = "Eslestirilmis: son - on")
  for (grup in names(basliklar)) {
    tablo <- sonuc[sonuc$degisken == grup, ]
    ozet <- setNames(tablo$deger, tablo$olcu)
    aralik <- ozet[c("alt_sinir", "ust_sinir")]
    pay <- .35 * ozet["hata_payi"]
    plot(NA, xlim = range(c(0, aralik)) + c(-pay, pay), ylim = c(-.5, .6), yaxt = "n", ylab = "",
         xlab = "Evren ortalama farki (puan)",
         main = paste(basliklar[grup], sprintf("p = %.6f", ozet["p_cift"]), sep = "\n"))
    segments(aralik[1], 0, aralik[2], 0, col = "#087F5B", lwd = 2)
    points(mean(aralik), 0, col = "#087F5B", pch = 19)
    abline(v = 0, lty = 2, col = "#555555")
    for (uc in aralik) {
      segments(uc, -.06, uc, .06, col = "#087F5B")
      text(uc, .15, sprintf("%.4f", uc))
    }
    legend("bottomright", legend = paste("Guven duzeyi:", 1 - veri$alfa), bty = "n", cex = .8)
  }
  cat("Grafik: ciktilar/r/fark-araliklari.png\n")
}

secenek <- commandArgs(trailingOnly = TRUE)
if (any(!secenek %in% c("--check", "--grafik"))) stop("Bilinmeyen secenek.")
veri <- read.csv("veri.csv", check.names = FALSE)
sonuc <- hesapla(veri)
print(sonuc, row.names = FALSE)
if ("--check" %in% secenek) kontrol_et(sonuc, read.csv("beklenen-sonuclar.csv", stringsAsFactors = FALSE))
if ("--grafik" %in% secenek) grafik_kaydet(veri)
sessionInfo()

File: companion/bolumler/b11/ornek-01/cozum.py
import argparse
import math
from pathlib import Path
import platform

import numpy as np
import pandas as pd
import scipy
from scipy import stats


SUTUNLAR = ["hacim_1", "ortalama_1", "standart_sapma_1", "hacim_2", "ortalama_2",
            "standart_sapma_2", "cift_sayisi", "ortalama_fark", "fark_sapmasi", "alfa"]


def hazirla(veri):
    if list(veri.columns) != SUTUNLAR or len(veri) != 1:
        raise ValueError("Tek ozet satiri ve dogru sutun sirasi gerekli.")
    if veri.isna().any().any():
        raise ValueError("Eksik ozet kabul edilmez.")
    for sutun in SUTUNLAR:
        if not pd.api.types.is_numeric_dtype(veri[sutun]) or pd.api.types.is_bool_dtype(veri[sutun]):
            raise ValueError("Sayisal ozet gerekli.")
    if not np.isfinite(veri.to_numpy(dtype=float)).all():
        raise ValueError("Sonlu ozet gerekli.")
    temiz = veri.reset_index(drop=True).copy(deep=True)
    for sutun in ["hacim_1", "hacim_2", "cift_sayisi"]:
        hacim = float(temiz.loc[0, sutun])
        if hacim < 2 or not hacim.is_integer():
            raise ValueError("Hacimler en az 2 olan tam sayilar olmali.")
    if any(temiz.loc[0, sutun] <= 0 for sutun in ["standart_sapma_1", "standart_sapma_2", "fark_sapmasi"]):
        raise ValueError("Bu paket pozitif standart sapmalar gerektirir.")
    if not 0 < temiz.loc[0, "alfa"] < 1:
        raise ValueError("Alfa 0 ile 1 arasinda olmali.")
    return temiz


def test_ozeti(fark, standart_hata, serbestlik, alfa):
    t_degeri = fark / standart_hata
    p_degeri = float(2 * stats.t.sf(abs(t_degeri), df=serbestlik))
    kritik = float(stats.t.ppf(1 - alfa / 2, df=serbestlik))
    hata_payi = kritik * standart_hata
    sonuc = dict(standart_hata=standart_hata, serbestlik=serbestlik, t=t_degeri,
                  p_cift=p_degeri, kritik_t=kritik, hata_payi=hata_payi,
                  alt_sinir=fark - hata_payi, ust_sinir=fark + hata_payi,
                  genislik=2 * hata_payi, reddet_cift=int(p_degeri < alfa))
    if not all(math.isfinite(value) for value in sonuc.values()):
        raise ValueError("Test sayisal olarak hesaplanamadi.")
    return sonuc


def hesapla(veri):
    temiz = hazirla(veri)
    deger = {key: float(temiz.loc[0, key]) for key in SUTUNLAR}
    katki_1 = deger["standart_sapma_1"] ** 2 / deger["hacim_1"]
    katki_2 = deger["standart_sapma_2"] ** 2 / deger["hacim_2"]
    toplam = katki_1 + katki_2
    serbestlik = toplam ** 2 / (katki_1 ** 2 / (deger["hacim_1"] - 1) +
                                katki_2 ** 2 / (deger["hacim_2"] - 1))
    fark = deger["ortalama_1"] - deger["ortalama_2"]
    welch = dict(katki_1=katki_1, katki_2=katki_2, fark=fark)
    welch.update(test_ozeti(fark, math.sqrt(toplam), serbestlik, deger["alfa"]))
    eslesmis = test_ozeti(deger["ortalama_fark"], deger["fark_sapmasi"] / math.sqrt(deger["cift_sayisi"]),
                          deger["cift_sayisi"] - 1, deger["alfa"])
    eslesmis["dz"] = deger["ortalama_fark"] / deger["fark_sapmasi"]
    rows = [("girdi", key, value) for key, value in deger.items()]
    rows += [("welch", key, value) for key, value in welch.items()]
    rows += [("eslesmis", key, value) for key, value in eslesmis.items()]
    sonuc = pd.DataFrame(rows, columns=["degisken", "olcu", "deger"])
    if not np.isfinite(sonuc.deger).all():
        raise ValueError("Sonlu sonuc hesaplanamadi.")
    return sonuc


def kontrol_et(sonuc, beklenen):
    if list(beklenen.columns) != ["degisken", "olcu", "deger"]:
        raise ValueError("Kontrol sutunlari uyusmuyor.")
    if not sonuc[["degisken", "olcu"]].equals(beklenen[["degisken", "olcu"]]):
        raise ValueError("Kontrol etiketleri veya sirasi uyusmuyor.")
    hedefler = pd.to_numeric(beklenen.deger, errors="raise")
    if not np.isfinite(hedefler).all() or not np.isfinite(sonuc.deger).all():
        raise ValueError("Sonlu kontrol degerleri gerekli.")
    if not all(math.isclose(actual, expected, rel_tol=1e-9, abs_tol=1e-9)
               for actual, expected in zip(sonuc.deger, hedefler)):
        raise ValueError("Kontrol degerleri uyusmuyor.")
    print(f"DOGRULANDI: {len(sonuc)} kontrol degeri eslesiyor.")


def grafik_kaydet(veri):
    import matplotlib
    matplotlib.use("Agg")
    import matplotlib.pyplot as plt

    temiz = hazirla(veri)
    sonuc = hesapla(temiz)
    fig, paneller = plt.subplots(1, 2, figsize=(10, 4))
    bilgiler = [("welch", "Welch: Grup 1 - Grup 2"), ("eslesmis", "Eslestirilmis: son - on")]
    for panel, (grup, baslik) in zip(paneller, bilgiler):
        ozet = sonuc.loc[sonuc.degisken == grup].set_index("olcu").deger
        merkez = (ozet["alt_sinir"] + ozet["ust_sinir"]) / 2
        panel.errorbar(merkez, 0, xerr=ozet["hata_payi"], fmt="o", capsize=7, color="#087F5B")
        panel.axvline(0, linestyle="--", color="#555555", label="H0: fark = 0")
        for uc in [ozet["alt_sinir"], ozet["ust_sinir"]]:
            panel.annotate(f"{uc:.4f}", (uc, 0), xytext=(0, 12), textcoords="offset points", ha="center")
        pay = .35 * ozet["hata_payi"]
        panel.set(xlim=(min(0, ozet["alt_sinir"]) - pay, max(0, ozet["ust_sinir"]) + pay),
                  ylim=(-.5, .6), yticks=[], xlabel="Evren ortalama farki (puan)",
                  title=f"{baslik}\np = {ozet['p_cift']:.6f}")
        panel.legend(loc="lower right", fontsize="small")
    fig.suptitle(f"Iki ayri ornek: %{100 * (1 - float(temiz.loc[0, 'alfa'])):g} fark araliklari")
    fig.tight_layout()
    hedef = Path("ciktilar/python")
    hedef.mkdir(parents=True, exist_ok=True)
    fig.savefig(hedef / "fark-araliklari.png", dpi=150)
    plt.close(fig)
    print("Grafik: ciktilar/python/fark-araliklari.png")


def main():
    parser = argparse.ArgumentParser(description="B11: ozetlerden Welch ve eslestirilmis t testi")
    parser.add_argument("--check", action="store_true")
    parser.add_argument("--grafik", action="store_true")
    secenek = parser.parse_args()
    veri = pd.read_csv("veri.csv")
    sonuc = hesapla(veri)
    print(sonuc.to_string(index=False))
    if secenek.check:
        kontrol_et(sonuc, pd.read_csv("beklenen-sonuclar.csv"))
    if secenek.grafik:
        grafik_kaydet(veri)
    print(f"Python {platform.python_version()}, NumPy {np.__version__}, pandas {pd.__version__}, SciPy {scipy.__version__}")


if __name__ == "__main__":
    main()